working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+
+### !end-callout
+
+### Planning the `validate_book` Refactor
+
+Our steps to plan a refactor with Copilot stay the same: we first need to identify our dependencies, then check if we have tests for code that will be affected. We will be refactoring the `validate_book` function in the `routes.py` file, so we should search the project for where the function is used to remind ourselves of our dependencies. 
+
+This is a scenario where Copilot shouldn't be used, since Copilot has access to open files, but not all of the files in your project. Indeed, if we ask Copilot for help, it will warn us about this lack of access and give us suggestions on how to search the project for dependencies.
+
+<details>
+  <summary>
+    If you're curious, feel free to try asking Copilot to identify dependencies for <code>`validate_book`</code> or expand this section to see our prompt and what Copilot suggested.
+  </summary>
+
+  **Prompt:**
+  > What code in this project is dependent on the validate_book function?
+
+  **Copilot's Response:**
+  > As an AI, I don't have access to your entire project's codebase, so I can't directly identify all the places where `validate_book` function is used. However, typically, such a function would be used in routes that require a book ID as a parameter.
+  >  
+  > For example, you might have routes like `@books_bp.route("/<book_id>", methods=["GET", "PUT", "DELETE"])` where you would use `validate_book(book_id)` to ensure the provided book ID is valid before proceeding with the GET, PUT, or DELETE operation.
+  >  
+  > Please search your codebase for `validate_book` to find all its usages. In Visual Studio Code, you can do this by pressing `Ctrl+Shift+F` and typing `validate_book` in the search bar.
+</details>
+
+If we search our project, we have 4 dependencies: 
+- the `validate_book` function definition
+- usage in `read_one_book`
+- usage in `update_book`
+- usage in `delete_book`
+
+Peeking at our test suite in `test_routes.py`, we do have tests for `read_one_book`, but we need to write tests for the remaining untested dependencies: `validate_book`, `update_book`, and `delete_book`. 
+
+### Updating Our Test Suite
+
+#### Testing `validate_book`
+
+We'll start with testing the `validate_book` function. Before we begin, we're going to open several files to ensure Copilot has plenty of context around our code and test set up. Take a moment to ensure the following files are open in VS Code:
+- `book.py`
+- `routes.py`
+- `conftest.py`
+- `test_models.py`
+- `test_routes.py`
+
+We're going to start from `routes.py` and use our cursor to highlight the entire `validate_book` function. Next, bring up the inline Copilot chat and invoke the `/tests` shortcut. When Copilot is done thinking, we may see a slightly different UI than in previous scenarios:
```

Comment:
- we can say what to do directly
- from would sound more natural if the passage spoke about working through several files starting from routes.py
```suggestion
Start in `routes.py` and use the cursor to highlight the entire `validate_book` function. Next, bring up the inline Copilot chat and invoke the `/tests` shortcut. When Copilot is done thinking, we may see a slightly different UI than in previous scenarios:
```